src/conn.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    marker::PhantomData,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use metrics::counter;
+use tokio::sync::OwnedSemaphorePermit;
+
+use crate::metrics_utils::{
+    GaugeGuard, ACTIVE_CONNECTIONS, CLOSED_TOTAL, IDLE_CONNECTIONS, OPENED_TOTAL, OPEN_CONNECTIONS,
+};
+
+pub(crate) struct ActiveConn<C> {
+    inner: C,
+    state: ConnState,
+    _permit: OwnedSemaphorePermit,
+    _active_connections_gauge: GaugeGuard,
+}
+
+impl<C> ActiveConn<C> {
+    pub(crate) fn new(inner: C, permit: OwnedSemaphorePermit, state: ConnState) -> ActiveConn<C> {
+        Self {
+            inner,
+            state,
+            _permit: permit,
```

Comment:
Why do you keep the permit instead of forgetting it? I can't see you using it anywhere?